src/js/heuristicblocking.js

```diff
@@ -104,33 +113,143 @@ HeuristicBlocker.prototype = {
     let fqdn = (new URI(details.url)).host,
       origin = window.getBaseDomain(fqdn);
 
-    // if this is a main window request
+    // if this is a main window request, update tab data and quit
     if (details.type == "main_frame") {
-      // save the origin associated with the tab
-      log("Origin: " + origin + "\tURL: " + details.url);
       tabOrigins[details.tabId] = origin;
+      tabURLs[details.tabId] = details.url;
       return {};
     }
 
-    let tabOrigin = tabOrigins[details.tabId];
+    let tabOrigin = tabOrigins[details.tabId],
+      self = this;
 
     // ignore first-party requests
     if (!tabOrigin || origin == tabOrigin) {
       return {};
     }
 
     // abort if we already made a decision for this FQDN
-    let action = this.storage.getAction(fqdn);
+    let action = self.storage.getAction(fqdn);
     if (action != constants.NO_TRACKING && action != constants.ALLOW) {
       return {};
     }
 
-    // ignore if there are no tracking cookies
-    if (!hasCookieTracking(details, origin)) {
+    // check if there are tracking cookies
+    if (hasCookieTracking(details, origin)) {
+      self._recordPrevalence(fqdn, origin, tabOrigin);
       return {};
     }
 
-    this._recordPrevalence(fqdn, origin, tabOrigin);
+    if (checkForCookieShare && details.type == 'image') {
+      // get all cookies for the top-level frame and pass those to the
+      // cookie-share accounting function
+      chrome.cookies.getAll({
+        url: tabURLs[details.tabId]
+      }, function(cookies) {
+        if (cookies.length >= 1) {
+          self.pixelCookieShareAccounting(details, cookies);
+        }
+      });
+    }
+  },
+
+  /**
+   * Checks for cookie sharing: requests to third-party domains that include
+   * high entropy data from first-party cookies (associated with the top-level
+   * frame). Only catches plain-text verbatim sharing (b64 encoding + the like
+   * defeat it). Assumes any long string that doesn't contain URL fragments or
+   * stopwords is an identifier.  Doesn't catch cookie syncing (3rd party -> 3rd
+   * party), but most of those tracking cookies should be blocked anyway.
+   *
+   * @param details are those from onBeforeSendHeaders
+   * @param cookies are the result of chrome.cookies.getAll()
+   * @returns {*}
+   */
+  pixelCookieShareAccounting: function (details, cookies) {
+    let fqdn = (new URI(details.url)).host,
+      origin = window.getBaseDomain(fqdn),
+      tabUrl = tabURLs[details.tabId],
+      tabOrigin = tabOrigins[details.tabId],
+      args = _extractArgs(details),
```

Comment:
I don't think we need `extractArgs`. Can avoid creating the intermediate `args` object and simply work with the `searchParams` iterator. Something like:
```javascript
...
searchParams = (new URL(response_url)).searchParams,
...
```